File: back_end/crossword_builder_api/utilities/character.py
from __future__ import annotations
from typing import Annotated, Literal
from pydantic import StringConstraints


### Types

Letter = Annotated[
    str,
    StringConstraints(
        min_length=1,
        max_length=1,
        pattern="^[A-Z]$"
    )
]
EmptySquareType = Literal["□"]
FilledSquareType = Literal["■"]
SuggestableCharacter = Letter | FilledSquareType
PatternCharacter = Letter | EmptySquareType
Character = Letter | FilledSquareType | EmptySquareType


### Values

def build_letters() -> list[Letter]:
    character_code_for_A = ord("A")
    return [
        chr(character_code_for_A + index)
        for index in range(26)
    ]

EMPTY_SQUARE = "□"
FILLED_SQUARE = "■"

File: front_end/src/utilities/character.ts
export type Letter = (
  'A' | 'B' | 'C' | 'D' | 'E' | 'F' | 'G' | 'H' | 'I' | 'J' | 'K' | 'L' | 'M' |
  'N' | 'O' | 'P' | 'Q' | 'R' | 'S' | 'T' | 'U' | 'V' | 'W' | 'X' | 'Y' | 'Z'
);
export type SuggestableCharacter = Letter | '■';
export type Character = Letter | '□' | '■';


export const LETTERS = [
  'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
  'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'
] as const;
export const EMPTY_SQUARE = '□';
export const FILLED_SQUARE = '■';


export function isLetter(
  value: any
): boolean {
  if (typeof value !== 'string') return false;
  return /^([A-Z])$/.test(value)
}

export function isSuggestableCharacter(
  value: any
): boolean {
  if (typeof value !== 'string') return false;
  return /^([A-Z]|■)$/.test(value);
}

export function isCharacter(
  value: any
): boolean {
  if (typeof value !== 'string') return false;
  return /^([A-Z]|■|□)$/.test(value);
}
